File: leetcode-submissions/0543-diameter-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        max_diameter = float("-inf")
        def dfs(node):

            if not node:
                return 0
            
            l =  dfs(node.left)
            r = dfs(node.right)

            nonlocal max_diameter
            max_diameter = max(max_diameter, l + r)
            return 1 + max(l, r)
        
        dfs(root)
        return max_diameter

File: leetcode-submissions/0543-diameter-of-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans;
    public int diameterOfBinaryTree(TreeNode root) {
        ans = 0;
        helper(root);
        return ans;
    }

    private int helper(TreeNode node) {
        if (node == null) {return 0;}
        int right =  helper(node.right);
        int left = helper(node.left);
        ans = Math.max(ans, left + right);
        return Math.max(right, left) + 1;
    }
}
